src/USER-MDI/fix_mdi_engine.cpp

```diff
@@ -0,0 +1,932 @@
+    forces[iforce] = 0.0;
+  }
+
+  // if not at a node, calculate the forces
+  if ( strcmp(current_node, "@DEFAULT") == 0 ) {
+    // certain fixes, such as shake, move the coordinates
+    // to ensure that the coordinates do not change, store a copy
+    double **x = atom->x;
+    for (int i = 0; i < nlocal; i++) {
+      x_buf[3*i+0] = x[i][0];
+      x_buf[3*i+1] = x[i][1];
+      x_buf[3*i+2] = x[i][2];
+    }
+
+    // calculate the forces
+    update->whichflag = 1; // 1 for dynamics
+    update->nsteps = 1;
+    lmp->init();
+    update->integrate->setup_minimal(1);
+  }
+
+  // pick local atoms from the buffer
+  double **f = atom->f;
+  for (int i = 0; i < nlocal; i++) {
+    forces[3*(atom->tag[i]-1)+0] = f[i][0]*forceconv;
+    forces[3*(atom->tag[i]-1)+1] = f[i][1]*forceconv;
+    forces[3*(atom->tag[i]-1)+2] = f[i][2]*forceconv;
+  }
+
+  // reduce the forces onto rank 0
+  MPI_Reduce(forces, forces_reduced, 3*atom->natoms, MPI_DOUBLE, MPI_SUM, 0, world);
+
+  // send the forces through MDI
+  if (master) {
+    ierr = MDI_Send((char*) forces_reduced, 3*atom->natoms, MDI_DOUBLE, driver_socket);
+    if (ierr != 0)
+      error->all(FLERR,"Unable to send atom forces to driver");
+  }
+
+  if ( strcmp(current_node, "@DEFAULT") == 0 ) {
+    // restore the original set of coordinates
+    double **x_new = atom->x;
+    for (int i = 0; i < nlocal; i++) {
+      x_new[i][0] = x_buf[3*i+0];
+      x_new[i][1] = x_buf[3*i+1];
+      x_new[i][2] = x_buf[3*i+2];
+    }
+  }
+
+  delete [] forces;
+  delete [] forces_reduced;
+  delete [] x_buf;
+
+}
+
+
+void FixMDIEngine::receive_forces(Error* error)
+{
+  double angstrom_to_bohr;
+  MDI_Conversion_Factor("angstrom", "bohr", &angstrom_to_bohr);
+  double kelvin_to_hartree;
+  MDI_Conversion_Factor("kelvin_energy", "hartree", &kelvin_to_hartree);
+
+  double potconv, posconv, forceconv;
+  potconv=kelvin_to_hartree/force->boltz;
+  posconv=force->angstrom/angstrom_to_bohr;
+  forceconv=potconv*posconv;
+
+  double *forces;
+  forces = new double[3*atom->natoms];
+
+  if (master) {
+    ierr = MDI_Recv((char*) forces, 3*atom->natoms, MDI_DOUBLE, driver_socket);
+    if (ierr != 0)
+      error->all(FLERR,"Unable to receive atom forces to driver");
+  }
+  MPI_Bcast(forces,3*atom->natoms,MPI_DOUBLE,0,world);
+
+  // pick local atoms from the buffer
+  double **f = atom->f;
+  int *mask = atom->mask;
+  int nlocal = atom->nlocal;
+
+  for (int i = 0; i < nlocal; i++) {
+    f[i][0] = forces[3*(atom->tag[i]-1)+0]/forceconv;
+    f[i][1] = forces[3*(atom->tag[i]-1)+1]/forceconv;
+    f[i][2] = forces[3*(atom->tag[i]-1)+2]/forceconv;
+  }
+
+  delete [] forces;
+}
+
+
+void FixMDIEngine::add_forces(Error* error)
+{
+  double angstrom_to_bohr;
+  MDI_Conversion_Factor("angstrom", "bohr", &angstrom_to_bohr);
+  double kelvin_to_hartree;
+  MDI_Conversion_Factor("kelvin_energy", "hartree", &kelvin_to_hartree);
+
+  double potconv, posconv, forceconv;
```

Comment:
This will likely fail with reduced units. Is there a test the stops this fix from being used with `units lj`?